edge-hub/src/Microsoft.Azure.Devices.Routing.Core/Evaluator.cs

```diff
@@ -53,9 +51,9 @@ public ISet<Route> Routes
             }
         }
 
-        public ISet<Endpoint> Evaluate(IMessage message)
+        public ISet<RouteResult> Evaluate(IMessage message)
         {
-            var endpoints = new HashSet<Endpoint>();
+            var results = new HashSet<RouteResult>();
```

Comment:
Actually this whole block can be done via single linq chain. I prefer the declarative approach to work with collections in c#
```csharp
// Because we are only reading here, it doesn't matter that it is under a lock
// ReSharper disable once InconsistentlySynchronizedField - compiledRoutes is immutable
ImmutableDictionary<string, CompiledRoute> snapshot = this.compiledRoutes;

// Multiple routes for the same endpoint can exist, in which case
// the message should be routed with the highest available priority
var routes = snapshot.Values
    .Where(cr => cr.Route.Source.Match(message.MessageSource) && EvaluateInternal(cr, message))
    .Select(cr => new RouteResult(cr.Route.Endpoint, cr.Route.Priority, cr.Route.TimeToLiveSecs))
    .GroupBy(r => r.Endpoint)
    .Select(dups => dups.OrderBy(r => r.Priority).First());

var results = new HashSet<RouteResult>(routes);
```

Reply:
my 2 cents: use ToHashSet method if above impl is used.

Reply:
Cool. I didn't know we have this method!

Reply:
neat! =)

---
In reply to: [376150004](https://github.com/Azure/iotedge/pull/2469#discussion_r376150004) [](ancestors = 376150004)